multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/UploadAppAsyncExecution.java

```diff
@@ -0,0 +1,289 @@
+                                                 getMonitorUploadStatusCallback(applicationToUploadContext.getApplication(),
+                                                                                extractedModulePath,
+                                                                                applicationToUploadContext.getStepLogger(),
+                                                                                applicationToUploadContext.getCorrelationId(),
+                                                                                applicationToUploadContext.getTaskId()),
+                                                 null);
+        } catch (Exception e) {
+            FileUtils.cleanUp(extractedModulePath, LOGGER);
+            throw new SLException(e,
+                                  Messages.ERROR_WHILE_STARTING_ASYNC_UPLOAD_OF_APP_WITH_NAME_0,
+                                  applicationToUploadContext.getApplication()
+                                                            .getName());
+        }
+    }
+
+    private AsyncExecutionState processCloudPackage(ProcessContext context, CloudControllerClient client, CloudPackage cloudPackage) {
+        CloudApplication cloudApp = client.getApplication(context.getVariable(Variables.APP_TO_PROCESS)
+                                                                 .getName());
+        if (context.getVariable(Variables.SHOULD_UPDATE_APPLICATION_DIGEST)) {
+            var appEnv = client.getApplicationEnvironment(cloudApp.getGuid());
+            String newApplicationDigest = context.getVariable(Variables.CALCULATED_APPLICATION_DIGEST);
+            attemptToUpdateApplicationDigest(client, cloudApp, appEnv, newApplicationDigest);
+        }
+        context.setVariable(Variables.CLOUD_PACKAGE, cloudPackage);
+        context.setVariable(Variables.APP_CONTENT_CHANGED, true);
+        return AsyncExecutionState.FINISHED;
+    }
+
+    private void attemptToUpdateApplicationDigest(CloudControllerClient client, CloudApplication app, Map<String, String> appEnv,
+                                                  String newApplicationDigest) {
+        new ApplicationEnvironmentUpdater(app, appEnv, client).updateApplicationEnvironment(Constants.ENV_DEPLOY_ATTRIBUTES,
+                                                                                            Constants.ATTR_APP_CONTENT_DIGEST,
+                                                                                            newApplicationDigest);
+    }
+
+    @Override
+    public String getPollingErrorMessage(ProcessContext context) {
+        CloudApplicationExtended appToProcess = context.getVariable(Variables.APP_TO_PROCESS);
+        return MessageFormat.format(Messages.ERROR_OCCURRED_DURING_APPLICATION_UPLOAD_0, appToProcess.getName());
+    }
+
+    MonitorUploadStatusCallback getMonitorUploadStatusCallback(CloudApplication app, Path file, StepLogger stepLogger, String correlationId,
```

Comment:
does this need to be an internal class?

Reply:
it was that way before and uses methods from the UploadStep